### ISSUE TITLE
Allow rust-analyzer to detect and perform type inference on items created by derive proc-macros

### ISSUE THREAD

**Issue by escritorio-gustavo** [rust-lang/rust-analyzer#16056]
I'm starting to mess around with proc macros. I created a derive macro for my structs to auto implement the type state builder pattern. It does that by creating a builder struct, and a bunch of type state thelpers, which are then fed into the associated type of my trait `impl` block.

It's working pretty well, but none of the structs created by the derive macro are picked up by intellisense (auto complete doesn't work, inlay hints show `!`) even though the code compiles and works as expected.

**Comment by lnicola** [comment]
Proc macros work in general, obviously, so there must be something specific to your project or setup:

![image](https://github.com/rust-lang/rust-analyzer/assets/308347/66e40c51-0255-4aa1-84e9-56d2922c359f)

Could be related to https://github.com/rust-lang/rust-analyzer/issues/15033.

**Comment by escritorio-gustavo** [comment]
Yes, the proc-macro is working perfectly, the emmited code is correct, but I get no intellisense at all. I believe it might be because rust-analyzer assumes a derive macro only declares trait impl blocks and thus doesn't check any structs created by them. But I really have no idea.

**Comment by escritorio-gustavo** [comment]
This is weird, I have commented out the generics from the offending struct and even restarted rust-analyzer's server and the issue persists, even though all the generated code (with and without the generics) is correct and working

**Comment by escritorio-gustavo** [comment]
Interesting, running this command does work, but the expansion stops at the closing bracket of `mod type_state`

**Comment by escritorio-gustavo** [comment]
Which means the ´mod buid´ and `impl`  blocks are completely missing as far as rust analyzer knows

**Comment by escritorio-gustavo** [comment]
This is very different from the output from `cargo expand`, which correctly shows the entire expansion
